flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/assigner/EventTimeAlignedAssigner.java

```diff
@@ -0,0 +1,301 @@
+ *  * regarding copyright ownership.  The ASF licenses this file
+ *  * to you under the Apache License, Version 2.0 (the
+ *  * "License"); you may not use this file except in compliance
+ *  * with the License.  You may obtain a copy of the License at
+ *  *
+ *  *   http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing,
+ *  * software distributed under the License is distributed on an
+ *  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  * KIND, either express or implied.  See the License for the
+ *  * specific language governing permissions and limitations
+ *  * under the License.
+ *
+ */
+
+package org.apache.iceberg.flink.source.assigner;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.flink.api.common.eventtime.TimestampAssigner;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitState;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.flink.calcite.shaded.com.google.common.collect.Iterables.getFirst;
+
+public class EventTimeAlignedAssigner implements SplitAssigner, WatermarkTracker.Listener {
+  private static final Logger log = LoggerFactory.getLogger(EventTimeAlignedAssigner.class);
+  private final WatermarkTracker watermarkTracker;
+  private final State state;
+  private final TimestampAssigner<IcebergSourceSplit> timestampAssigner;
+  private final FutureNotifier futureNotifier;
+  private final Options options;
+
+  // flag that informs the state that there are no more splits to be added by the Enumerator
+  private volatile boolean noMoreSplits;
+
+  public EventTimeAlignedAssigner(
+      WatermarkTracker watermarkTracker,
+      TimestampAssigner<IcebergSourceSplit> timestampAssigner,
+      Options options) {
+    this.watermarkTracker = watermarkTracker;
+    this.state = new State(watermarkTracker, timestampAssigner);
+    this.options = options;
+    this.timestampAssigner = timestampAssigner;
+    this.futureNotifier = new FutureNotifier();
+    this.watermarkTracker.onInitialization();
+  }
+
+  @Override
+  public void start() {
+    watermarkTracker.addListener(this);
+  }
+
+  @Override
+  public void close() {
+    watermarkTracker.removeListener(this);
+    notifyListener();
+  }
+
+  @Override
+  public GetSplitResult getNext(@Nullable String hostname) {
+    GetSplitResult result = getNextInternal();
+    if (result.status() == GetSplitResult.Status.AVAILABLE) {
```

Comment:
What happens if the status is different?
Is this an error? Why do we return it then?